src/khepri_machine.erl

```diff
@@ -1020,63 +1028,68 @@ add_applied_condition1(StoreId, Options, Timeout) ->
-        Other when Other =/= false ->
-            Other
+            add_applied_condition2(StoreId, Options, Timeout);
+        false ->
+            T0 = khepri_utils:start_timeout_window(Timeout),
+            QueryFun = fun erlang:is_tuple/1,
+            case process_query1(StoreId, QueryFun, Timeout) of
+                true ->
+                    NewTimeout = khepri_utils:end_timeout_window(Timeout, T0),
+                    add_applied_condition2(StoreId, Options, NewTimeout);
+                Other when Other =/= false ->
+                    Other
+            end
     end.
 
 add_applied_condition2(StoreId, Options, Timeout) ->
-    %% After the previous local query, there is a great chance that the leader
-    %% was cached, though not 100% guarantied.
+    %% After the previous local query or sync command if there was one, there
+    %% is a great chance that the leader was cached, though not 100%
+    %% guarantied.
     case ra_leaderboard:lookup_leader(StoreId) of
         LeaderId when LeaderId =/= undefined ->
             add_applied_condition3(StoreId, Options, LeaderId, Timeout);
         undefined ->
+            %% Let's consider that the previous message was async. This will
+            %% force a new local query during the retry. We should know the
+            %% leader after that.
+            last_message_was_async(StoreId),
             add_applied_condition1(StoreId, Options, Timeout)
     end.
 
 add_applied_condition3(StoreId, Options, LeaderId, Timeout) ->
-    %% We query the leader to know the last index it committed. We also
-    %% double-check it is still the leader; if it is not, we recurse.
+    %% We query the leader to know the last index it committed in which term.
     T0 = khepri_utils:start_timeout_window(Timeout),
-    case ra:member_overview(LeaderId, Timeout) of
-        {ok, Overview, LeaderId} ->
-            NewTimeout = khepri_utils:end_timeout_window(Timeout, T0),
-
-            %% Now that we know the last committed index of the leader, we can
-            %% perform an arbitrary query on the local server. The query will
-            %% wait for that same index to be applied locally before it is
-            %% executed.
-            %%
-            %% We don't care about the result of that query. We just want to
-            %% block until the latest commands are applied locally.
-            #{log := #{last_index := LastIndex},
-              current_term := CurrentTerm} = Overview,
-            Condition = {applied, {LastIndex, CurrentTerm}},
-            Options1 = Options#{condition => Condition,
-                                timeout => NewTimeout},
-            {ok, Options1};
-        {ok, _Overview, NewLeaderId} ->
-            NewTimeout = khepri_utils:end_timeout_window(Timeout, T0),
-            add_applied_condition3(StoreId, Options, NewLeaderId, NewTimeout);
-        {timeout, _LeaderId} ->
+    try
+        case ra:key_metrics(LeaderId, Timeout) of
```

Comment:
I believe the syntax for try + case can be combined:

```erl
try ra:key_metrics(LeaderId, Timeout) of
    #{last_index := LastIndex, ..} ->
        %% ...
    _ ->
        %% ...
catch
    error:{erpc, timeout} ->
        %% ...
end
```

It might cut down on nesting a bit. (Not something I feel strongly about though if you prefer the separate try + case)

Reply:
Good idea, I will do that.